src/crystal/scheduler.cr

```diff
@@ -57,9 +57,24 @@ class Crystal::Scheduler
   end
 
   protected def resume(fiber : Fiber) : Nil
+    validate_resumable(fiber)
```

Comment:
Should we have another state and validate that the `fiber` is `resuming`? Otherwise we are not enforcing that the `fiber` is not trying to be resumed from two threads at the same time.

Maybe the flag is not just a boolean but a pointer to the thread so we can set & validate the flag is still owned by the current thread. Maybe I am missing something but although there are not a lot of code between the validation and the swapcontext, there is still a chance for other thread to try to do the same.

Reply:
Yes, I assume the following, otherwise this is still MT unsafe:
1. a Fiber must only be queued once;
2. a Scheduler's queue must have atomic operations so a Fiber may only dequeued (including stolen) once.

Since Fiber isn't a public API, and developers should use provided synchronization primitives (e.g. Channel, Mutex, select), I believe these assumptions to be acceptable. Assumption 1 is the whole reason I disabled `select` in my MT branch, by the way —I'll eventually reenable it by spawning distinct fibers for each wait operation.

Reply:
Wouldn't be safer and sound to add this check I propose then? For the right usages, there should be almost no overhead and will prevent invalid state. For me, it sounds like the same you wanted to avoid by rescheduling a dead fiber.

Reply:
You're right, it's safer to add another check. It'll add an atomic op, to make sure one scheduler only gets the right to resume the fiber.

Reply:
I'm trying to add this extra check, or changing `resumable` to be an atomic, but that's really tricky, and I'm not sure how to implement it :/

Reply:
I was thinking of something like this: https://github.com/bcardiff/crystal/commit/ddf3fc18ff61a14e7b499ba837cd4852381d7438 

WDYT?

The following snippet stresses the check:

```crystal
f = Fiber.new do
  while true
  end
end

Thread.new do
  Crystal::Scheduler.resume(f) # Boom!
end

Crystal::Scheduler.resume(f) # or Boom!
```

Reply:
Thanks!

I found a hack in the meantime in my MT branch (add a cas on resumable to set it from 1 to 2), but didn't check that it did work yet.

Reply:
If you agree with that commit, do you want to add it to this PR then so we can get some ✅from others?

I tried to come up with a version using just ints to flag the state, but it was more complex to mark the lifecycle of the "incarnations" of each fiber and, at the end of the day, knowing which thread is actually executing/willing to execute each fiber is probably valuable information.

Reply:
(also, it needs a rebase on master so the CI can be happy ...)